Investigate locally failed simtools-derive-mirror-rnda_psf_measurement

```
======================================================================================== FAILURES =========================================================================================
_______________________________________________________ test_applications_from_config[simtools-derive-mirror-rnda_psf_measurement] ________________________________________________________
[gw4] linux -- Python 3.11.9 /workdir/env/bin/python3.11

tmp_test_directory = local('/tmp/pytest-of-root/pytest-1/popen-gw4/test-data2')
config = {'APPLICATION': 'simtools-derive-mirror-rnda', 'CONFIGURATION': {'CONTAINMENT_FRACTION': 0.8, 'LOG_LEVEL': 'DEBUG', 'M...ue', 'TEST_COLUMN_NAME': 'mirror_reflection_random_angle_sigma1'}], 'TOLERANCE': 0.3}], 'TEST_NAME': 'psf_measurement'}
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0xffff787ead50>
request = <FixtureRequest for <Function test_applications_from_config[simtools-derive-mirror-rnda_psf_measurement]>>, model_version = '6.0.0'

    @pytest.mark.parametrize("config", test_configs, ids=test_ids)
    def test_applications_from_config(tmp_test_directory, config, monkeypatch, request, model_version):
        """
        Test all applications from config files found in the config directory.
    
        Parameters
        ----------
        tmp_test_directory: str
            Temporary directory, into which test configuration and output is written.
        config: dict
            Dictionary with the configuration parameters for the test.
    
        """
    
        tmp_config = copy.deepcopy(config)
        helpers.skip_camera_efficiency(tmp_config)
    
        # The db_add_file_to_db.py application requires a user confirmation.
        # With this line we mock the user confirmation to be y for the test
        # Notice this is done for all tests, so keep in mind if in the future we add tests with input.
        monkeypatch.setattr("sys.stdin", StringIO("y\n"))
    
        logger.info(f"Test configuration from config file: {tmp_config}")
        logger.info(f"Model version: {request.config.getoption('--model_version')}")
        logger.info(f"Application configuration: {tmp_config}")
        cmd, config_file_model_version = configuration.configure(
            tmp_config, tmp_test_directory, request
        )
    
        logger.info(f"Running application: {cmd}")
        assert os.system(cmd) == 0, f"Application failed: {cmd}"
    
>       validate_output.validate_all_tests(tmp_config, request, config_file_model_version)

tests/integration_tests/test_applications_from_config.py:54: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
simtools/testing/validate_output.py:30: in validate_all_tests
    validate_application_output(config)
simtools/testing/validate_output.py:56: in validate_application_output
    _validate_reference_output_file(config, integration_test)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

config = {'APPLICATION': 'simtools-derive-mirror-rnda', 'CONFIGURATION': {'CONTAINMENT_FRACTION': 0.8, 'LOG_LEVEL': 'DEBUG', 'M...ue', 'TEST_COLUMN_NAME': 'mirror_reflection_random_angle_sigma1'}], 'TOLERANCE': 0.3}], 'TEST_NAME': 'psf_measurement'}
integration_test = {'REFERENCE_OUTPUT_FILE': 'tests/resources/derive_mirror_rnda_psf_measurement.ecsv', 'TEST_COLUMNS': [{'CUT_COLUMN_NAM...'best_fit', 'CUT_CONDITION': '==True', 'TEST_COLUMN_NAME': 'mirror_reflection_random_angle_sigma1'}], 'TOLERANCE': 0.3}

    def _validate_reference_output_file(config, integration_test):
        """Compare with reference output file."""
>       assert compare_files(
            integration_test["REFERENCE_OUTPUT_FILE"],
            Path(config["CONFIGURATION"]["OUTPUT_PATH"]).joinpath(
                config["CONFIGURATION"]["OUTPUT_FILE"]
            ),
            integration_test.get("TOLERANCE", 1.0e-5),
            integration_test.get("TEST_COLUMNS", None),
        )
E       AssertionError

simtools/testing/validate_output.py:72: AssertionError
---------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------

Measured D80:
Mean = 1.464 cm, StdDev = 0.083 cm

Simulated D80:
Mean = 1.056 cm, StdDev = 0.050 cm

mirror_random_reflection_angle
Previous value = 0.006300
New value = 0.005007
```

--- a/src/simtools/applications/derive_mirror_rnda.py
+++ b/src/simtools/applications/derive_mirror_rnda.py
@@ -179,6 +179,13 @@ def _parse(label):
         required=False,
     )
     config.parser.add_argument(
+        "--rtol_psf_containment",
+        help=("Relative tolerance for the containment diameter. " "Default is 0.1 (10%)."),
+        type=float,
+        required=False,
+        default=0.1,
+    )
+    config.parser.add_argument(
         "--use_random_focal_length",
         help=("Use random focal lengths."),
         action="store_true",

--- a/src/simtools/applications/derive_mirror_rnda.py
+++ b/src/simtools/applications/derive_mirror_rnda.py
@@ -180,7 +180,7 @@ def _parse(label):
     )
     config.parser.add_argument(
         "--rtol_psf_containment",
-        help=("Relative tolerance for the containment diameter. " "Default is 0.1 (10%)."),
+        help="Relative tolerance for the containment diameter (default is 0.1).",
         type=float,
         required=False,
         default=0.1,

--- a/src/simtools/ray_tracing/mirror_panel_psf.py
+++ b/src/simtools/ray_tracing/mirror_panel_psf.py
@@ -132,49 +132,74 @@ class MirrorPanelPSF:
         if self.args_dict["no_tuning"]:
             self.rnda_opt = self.rnda_start
         else:
-            self._optimize_reflection_angle()
+            self._optimize_reflection_angle(
+                relative_tolerance_d80=self.args_dict["rtol_psf_containment"],
+            )
 
         self.mean_d80, self.sig_d80 = self.run_simulations_and_analysis(
             self.rnda_opt, save_figures=save_figures
         )
 
-    def _optimize_reflection_angle(self, step_size=0.1):
-        """Optimize the random reflection angle to minimize the difference in D80 containment."""
+    def _optimize_reflection_angle(
+        self, relative_tolerance_d80=0.1, step_size=0.1, max_iteration=100
+    ):
+        """
+        Optimize the random reflection angle to minimize the difference in D80 containment.
+
+        Parameters
+        ----------
+        relative_tolerance_d80: float
+            Relative tolerance for the D80 containment.
+        step_size: float
+            Initial step size for optimization.
+        max_iteration: int
+            Maximum number of iterations.
+
+        Raises
+        ------
+        ValueError
+            If the optimization reaches the maximum number of iterations without converging.
+
+        """
+        self._logger.info(
+            "Optimizing random reflection angle "
+            f"(relative tolerance = {relative_tolerance_d80}, "
+            f"step size = {step_size}, max iteration = {max_iteration})"
+        )
 
         def collect_results(rnda, mean, sig):
             self.results_rnda.append(rnda)
             self.results_mean.append(mean)
             self.results_sig.append(sig)
 
-        stop = False
-        mean_d80, sig_d80 = self.run_simulations_and_analysis(self.rnda_start)
+        reference_d80 = self.args_dict["psf_measurement_containment_mean"]
         rnda = self.rnda_start
-        sign_delta = np.sign(mean_d80 - self.args_dict["psf_measurement_containment_mean"])
-        collect_results(rnda, mean_d80, sig_d80)
-        while not stop:
-            rnda = rnda - (step_size * self.rnda_start * sign_delta)
-            if rnda < 0:
-                rnda = 0
-                collect_results(rnda, mean_d80, sig_d80)
-                break
+        prev_error_d80 = float("inf")
+        iteration = 0
+
+        while True:
             mean_d80, sig_d80 = self.run_simulations_and_analysis(rnda)
-            new_sign_delta = np.sign(mean_d80 - self.args_dict["psf_measurement_containment_mean"])
-            stop = new_sign_delta != sign_delta
-            sign_delta = new_sign_delta
+            error_d80 = abs(1 - mean_d80 / reference_d80)
             collect_results(rnda, mean_d80, sig_d80)
 
-        self._interpolate_optimal_rnda()
+            if error_d80 < relative_tolerance_d80:
+                break
 
-    def _interpolate_optimal_rnda(self):
-        """Interpolate to find the optimal random reflection angle."""
-        self.results_rnda, self.results_mean, self.results_sig = gen.sort_arrays(
-            self.results_rnda, self.results_mean, self.results_sig
-        )
-        self.rnda_opt = np.interp(
-            x=self.args_dict["psf_measurement_containment_mean"],
-            xp=self.results_mean,
-            fp=self.results_rnda,
-        )
+            if mean_d80 < reference_d80:
+                rnda += step_size * self.rnda_start
+            else:
+                rnda -= step_size * self.rnda_start
+
+            if error_d80 >= prev_error_d80:
+                step_size = step_size / 2
+            prev_error_d80 = error_d80
+            iteration += 1
+            if iteration > max_iteration:
+                raise ValueError(
+                    f"Maximum iterations ({max_iteration}) reached without convergence."
+                )
+
+        self.rnda_opt = rnda
 
     def _get_starting_value(self):
         """Get optimization starting value from command line or previous model."""

--- a/src/simtools/ray_tracing/mirror_panel_psf.py
+++ b/src/simtools/ray_tracing/mirror_panel_psf.py
@@ -132,24 +132,18 @@ class MirrorPanelPSF:
         if self.args_dict["no_tuning"]:
             self.rnda_opt = self.rnda_start
         else:
-            self._optimize_reflection_angle(
-                relative_tolerance_d80=self.args_dict["rtol_psf_containment"],
-            )
+            self._optimize_reflection_angle()
 
         self.mean_d80, self.sig_d80 = self.run_simulations_and_analysis(
             self.rnda_opt, save_figures=save_figures
         )
 
-    def _optimize_reflection_angle(
-        self, relative_tolerance_d80=0.1, step_size=0.1, max_iteration=100
-    ):
+    def _optimize_reflection_angle(self, step_size=0.1, max_iteration=100):
         """
         Optimize the random reflection angle to minimize the difference in D80 containment.
 
         Parameters
         ----------
-        relative_tolerance_d80: float
-            Relative tolerance for the D80 containment.
         step_size: float
             Initial step size for optimization.
         max_iteration: int
@@ -161,6 +155,7 @@ class MirrorPanelPSF:
             If the optimization reaches the maximum number of iterations without converging.
 
         """
+        relative_tolerance_d80 = self.args_dict["rtol_psf_containment"]
         self._logger.info(
             "Optimizing random reflection angle "
             f"(relative tolerance = {relative_tolerance_d80}, "

--- a/src/simtools/testing/validate_output.py
+++ b/src/simtools/testing/validate_output.py
@@ -235,6 +235,7 @@ def compare_ecsv_files(file1, file2, tolerance=1.0e-5, test_columns=None):
 
         if np.issubdtype(table1_masked[col_name].dtype, np.floating):
             if not np.allclose(table1_masked[col_name], table2_masked[col_name], rtol=tolerance):
+                _logger.warning(f"Column {col_name} outside of relative tolerance {tolerance}")
                 return False
 
     return True

--- a/tests/integration_tests/config/derive_mirror_rnda_psf_measurement.yml
+++ b/tests/integration_tests/config/derive_mirror_rnda_psf_measurement.yml
@@ -9,6 +9,7 @@ CTA_SIMPIPE:
     PSF_MEASUREMENT: ./tests/resources/MLTdata-preproduction.ecsv
     USE_RANDOM_FOCAL_LENGTH: true
     RNDA: 0.0063
+    RTOL_PSF_CONTAINMENT: 0.1
     TEST: true
     OUTPUT_PATH: simtools-output
     OUTPUT_FILE: test_derive_mirror_rnda_psf_measurement.ecsv

--- a/tests/unit_tests/ray_tracing/test_mirror_panel_psf.py
+++ b/tests/unit_tests/ray_tracing/test_mirror_panel_psf.py
@@ -1,6 +1,7 @@
 #!/usr/bin/python3
 
 import copy
+import re
 from unittest.mock import patch
 
 import astropy.units as u
@@ -23,6 +24,7 @@ def mock_args_dict(tmp_test_directory):
         "random_focal_length": None,
         "no_tuning": False,
         "rnda": 0,
+        "rtol_psf_containment": 0.1,
         "simtel_path": "path/to/simtel",
         "number_of_mirrors_to_test": 2,
         "use_random_focal_length": False,
@@ -309,47 +311,33 @@ def test_derive_random_reflection_angle_with_tuning(
         assert mirror_psf.sig_d80 == 0.1
 
 
-def test_interpolate_optimal_rnda(mock_mirror_panel_psf):
-    mirror_psf = copy.deepcopy(mock_mirror_panel_psf)
-    mirror_psf.results_rnda = [0.1, 0.2, 0.3]
-    mirror_psf.results_mean = [0.4, 0.5, 0.6]
-    mirror_psf.results_sig = [0.01, 0.02, 0.03]
-    mirror_psf.args_dict["psf_measurement_containment_mean"] = 0.55
-
-    with patch("numpy.interp", return_value=0.25) as mock_interp:
-        mirror_psf._interpolate_optimal_rnda()
-
-        mock_interp.assert_called_once_with(
-            x=0.55,
-            xp=[0.4, 0.5, 0.6],
-            fp=[0.1, 0.2, 0.3],
-        )
-        assert mirror_psf.rnda_opt == 0.25
-
-
 def test_optimize_reflection_angle(mock_mirror_panel_psf, mock_run_simulations_and_analysis_string):
     mirror_psf = copy.deepcopy(mock_mirror_panel_psf)
     mirror_psf.rnda_start = 0.1
     mirror_psf.args_dict["psf_measurement_containment_mean"] = 0.5
 
-    with (
-        patch(
-            mock_run_simulations_and_analysis_string,
-            side_effect=[
-                (0.7, 0.1),  # First call
-                (0.6, 0.1),  # Second call
-                (0.4, 0.1),  # Third call
-            ],
-        ) as mock_run_simulations,
-        patch(
-            "simtools.ray_tracing.mirror_panel_psf.MirrorPanelPSF._interpolate_optimal_rnda"
-        ) as mock_interpolate,
-    ):
+    with patch(
+        mock_run_simulations_and_analysis_string,
+        side_effect=[
+            (0.7, 0.1),  # First call
+            (0.6, 0.1),  # Second call
+            (0.45, 0.1),  # Third call
+        ],
+    ) as mock_run_simulations:
 
-        mirror_psf._optimize_reflection_angle(0.9)
+        mirror_psf._optimize_reflection_angle(relative_tolerance_d80=0.1)
 
-        assert mirror_psf.results_rnda == pytest.approx([0.1, 0.01, 0.0])
-        assert mirror_psf.results_mean == pytest.approx([0.7, 0.6, 0.6])
+        assert mirror_psf.results_rnda == pytest.approx([0.1, 0.09, 0.08])
+        assert mirror_psf.results_mean == pytest.approx([0.7, 0.6, 0.45])
         assert mirror_psf.results_sig == pytest.approx([0.1, 0.1, 0.1])
-        mock_interpolate.assert_called_once()
-        assert mock_run_simulations.call_count == 2
+        assert mock_run_simulations.call_count == 3
+
+    with patch(
+        mock_run_simulations_and_analysis_string,
+        side_effect=lambda *args, **kwargs: (0.6, 0.1),
+    ) as mock_run_simulations:
+        with pytest.raises(
+            ValueError,
+            match=re.escape("Maximum iterations (100) reached without convergence."),
+        ):
+            mirror_psf._optimize_reflection_angle(relative_tolerance_d80=0.1)

--- a/tests/unit_tests/ray_tracing/test_mirror_panel_psf.py
+++ b/tests/unit_tests/ray_tracing/test_mirror_panel_psf.py
@@ -325,7 +325,7 @@ def test_optimize_reflection_angle(mock_mirror_panel_psf, mock_run_simulations_a
         ],
     ) as mock_run_simulations:
 
-        mirror_psf._optimize_reflection_angle(relative_tolerance_d80=0.1)
+        mirror_psf._optimize_reflection_angle()
 
         assert mirror_psf.results_rnda == pytest.approx([0.1, 0.09, 0.08])
         assert mirror_psf.results_mean == pytest.approx([0.7, 0.6, 0.45])
@@ -340,4 +340,4 @@ def test_optimize_reflection_angle(mock_mirror_panel_psf, mock_run_simulations_a
             ValueError,
             match=re.escape("Maximum iterations (100) reached without convergence."),
         ):
-            mirror_psf._optimize_reflection_angle(relative_tolerance_d80=0.1)
+            mirror_psf._optimize_reflection_angle()

